katdal/datasources.py

```diff
@@ -53,24 +151,73 @@ def name(self):
                 sensors[key] = TelstateSensorData(telstate, key)
-        metadata = AttrsSensors(attrs, sensors, name=source_name)
-        DataSource.__init__(self, metadata, None)
+        metadata = AttrsSensors(telstate, sensors, name=source_name)
+        try:
+            base_name = telstate['chunk_name']
+            chunk_info = telstate['chunk_info']
+        except KeyError:
+            # Metadata without data
+            DataSource.__init__(self, metadata, None)
+        else:
+            # Extract VisFlagsWeights and timestamps from telstate
+            store = RadosChunkStore.from_config(telstate['ceph_conf'],
+                                                telstate['ceph_pool'])
+            ts_name = store.join(base_name, 'timestamps')
+            ts_chunks = chunk_info['timestamps']['chunks']
+            ts_dtype = chunk_info['timestamps']['dtype']
+            timestamps = store.get_dask_array(ts_name, ts_chunks, ts_dtype)
+            timestamps = DaskLazyIndexer(timestamps)
+            data = ChunkStoreVisFlagsWeights(store, base_name, chunk_info)
+            DataSource.__init__(self, metadata, timestamps, data)
+
+    @classmethod
+    def from_url(cls, url):
+        """Construct TelstateDataSource from URL (RDB file / REDIS server)."""
+        url_parts = urlparse.urlparse(url, scheme='file')
+        kwargs = dict(urlparse.parse_qsl(url_parts.query))
+        kwargs['source_name'] = url_parts.geturl()
+        if url_parts.scheme == 'file':
+            # RDB dump file
+            telstate = katsdptelstate.TelescopeState()
+            try:
+                telstate.load_from_file(url_parts.path)
+            except OSError as err:
+                raise DataSourceNotFound(str(err))
+            return cls(telstate, **kwargs)
+        elif url_parts.scheme == 'redis':
+            # Redis server
+            try:
+                telstate = katsdptelstate.TelescopeState(url_parts.netloc)
+            except redis.exceptions.TimeoutError as err:
+                raise DataSourceNotFound(str(err))
+            return cls(telstate, **kwargs)
 
 
 def open_data_source(url):
     """Construct the data source described by the given URL."""
-    url_parts = urlparse.urlparse(url)
-    if url_parts.scheme == 'telstate+redis':
-        telstate = katsdptelstate.TelescopeState(url_parts.netloc)
-        return TelstateDataSource(telstate, url)
-    else:
-        raise ValueError("Unsupported data source '%s'" % (url,))
+    try:
+        return TelstateDataSource.from_url(url)
+    except DataSourceNotFound as err:
+        # Amend the error message for the case of an IP address without scheme
+        url_parts = urlparse.urlparse(url, scheme='file')
+        if url_parts.scheme == 'file' and not os.path.isfile(url_parts.path):
+            raise DataSourceNotFound(
+                '{} (add a URL scheme if it is not meant to be a file)'
+                .format(str(err), url_parts.path))
```

Comment:
You've got one `{}` and two arguments to format. Have you tested it?

Reply:
Oops, no. katdal has no unit tests, remember? (at least at this level - ChunkStore puts the rest to shame!)

Reply:
Fixed.